{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "reserved-professional",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "usage: /home/joewu/miniconda3/bin/freebayes [OPTION] ... [BAM FILE] ... \n",
      "\n",
      "Bayesian haplotype-based polymorphism discovery.\n",
      "\n",
      "citation: Erik Garrison, Gabor Marth\n",
      "          \"Haplotype-based variant detection from short-read sequencing\"\n",
      "          arXiv:1207.3907 (http://arxiv.org/abs/1207.3907)\n",
      "\n",
      "overview:\n",
      "\n",
      "    To call variants from aligned short-read sequencing data, supply BAM files and\n",
      "    a reference.  FreeBayes will provide VCF output on standard out describing SNPs,\n",
      "    indels, and complex variants in samples in the input alignments.\n",
      "\n",
      "    By default, FreeBayes will consider variants supported by at least 2\n",
      "    observations in a single sample (-C) and also by at least 20% of the reads from\n",
      "    a single sample (-F).  These settings are suitable to low to high depth\n",
      "    sequencing in haploid and diploid samples, but users working with polyploid or\n",
      "    pooled samples may wish to adjust them depending on the characteristics of\n",
      "    their sequencing data.\n",
      "\n",
      "    FreeBayes is capable of calling variant haplotypes shorter than a read length\n",
      "    where multiple polymorphisms segregate on the same read.  The maximum distance\n",
      "    between polymorphisms phased in this way is determined by the\n",
      "    --max-complex-gap, which defaults to 3bp.  In practice, this can comfortably be\n",
      "    set to half the read length.\n",
      "\n",
      "    Ploidy may be set to any level (-p), but by default all samples are assumed to\n",
      "    be diploid.  FreeBayes can model per-sample and per-region variation in\n",
      "    copy-number (-A) using a copy-number variation map.\n",
      "\n",
      "    FreeBayes can act as a frequency-based pooled caller and describe variants\n",
      "    and haplotypes in terms of observation frequency rather than called genotypes.\n",
      "    To do so, use --pooled-continuous and set input filters to a suitable level.\n",
      "    Allele observation counts will be described by AO and RO fields in the VCF output.\n",
      "\n",
      "\n",
      "examples:\n",
      "\n",
      "    # call variants assuming a diploid sample\n",
      "    freebayes -f ref.fa aln.bam >var.vcf\n",
      "\n",
      "    # require at least 5 supporting observations to consider a variant\n",
      "    freebayes -f ref.fa -C 5 aln.bam >var.vcf\n",
      "\n",
      "    # use a different ploidy\n",
      "    freebayes -f ref.fa -p 4 aln.bam >var.vcf\n",
      "\n",
      "    # assume a pooled sample with a known number of genome copies\n",
      "    freebayes -f ref.fa -p 20 --pooled-discrete aln.bam >var.vcf\n",
      "\n",
      "    # generate frequency-based calls for all variants passing input thresholds\n",
      "    freebayes -f ref.fa -F 0.01 -C 1 --pooled-continuous aln.bam >var.vcf\n",
      "\n",
      "    # use an input VCF (bgzipped + tabix indexed) to force calls at particular alleles\n",
      "    freebayes -f ref.fa -@ in.vcf.gz aln.bam >var.vcf\n",
      "\n",
      "    # generate long haplotype calls over known variants\n",
      "    freebayes -f ref.fa --haplotype-basis-alleles in.vcf.gz \\ \n",
      "                        --haplotype-length 50 aln.bam\n",
      "\n",
      "    # naive variant calling: simply annotate observation counts of SNPs and indels\n",
      "    freebayes -f ref.fa --haplotype-length 0 --min-alternate-count 1 \\ \n",
      "        --min-alternate-fraction 0 --pooled-continuous --report-monomorphic >var.vcf\n",
      "\n",
      "\n",
      "parameters:\n",
      "\n",
      "   -h --help       Prints this help dialog.\n",
      "   --version       Prints the release number and the git commit id.\n",
      "\n",
      "input and output:\n",
      "\n",
      "   -b --bam FILE   Add FILE to the set of BAM files to be analyzed.\n",
      "   -L --bam-list FILE\n",
      "                   A file containing a list of BAM files to be analyzed.\n",
      "   -c --stdin      Read BAM input on stdin.\n",
      "   -v --vcf FILE   Output VCF-format results to FILE.\n",
      "   -f --fasta-reference FILE\n",
      "                   Use FILE as the reference sequence for analysis.\n",
      "                   An index file (FILE.fai) will be created if none exists.\n",
      "                   If neither --targets nor --region are specified, FreeBayes\n",
      "                   will analyze every position in this reference.\n",
      "   -t --targets FILE\n",
      "                   Limit analysis to targets listed in the BED-format FILE.\n",
      "   -r --region <chrom>:<start_position>-<end_position>\n",
      "                   Limit analysis to the specified region, 0-base coordinates,\n",
      "                   end_position not included (same as BED format).\n",
      "                   Either '-' or '..' maybe used as a separator.\n",
      "   -s --samples FILE\n",
      "                   Limit analysis to samples listed (one per line) in the FILE.\n",
      "                   By default FreeBayes will analyze all samples in its input\n",
      "                   BAM files.\n",
      "   --populations FILE\n",
      "                   Each line of FILE should list a sample and a population which\n",
      "                   it is part of.  The population-based bayesian inference model\n",
      "                   will then be partitioned on the basis of the populations.\n",
      "   -A --cnv-map FILE\n",
      "                   Read a copy number map from the BED file FILE, which has\n",
      "                   the format:\n",
      "                      reference sequence, start, end, sample name, copy number\n",
      "                   ... for each region in each sample which does not have the\n",
      "                   default copy number as set by --ploidy.\n",
      "   --trace FILE    Output an algorithmic trace to FILE.\n",
      "   --failed-alleles FILE\n",
      "                   Write a BED file of the analyzed positions which do not\n",
      "                   pass --pvar to FILE.\n",
      "   -@ --variant-input VCF\n",
      "                   Use variants reported in VCF file as input to the algorithm.\n",
      "                   Variants in this file will included in the output even if\n",
      "                   there is not enough support in the data to pass input filters.\n",
      "   -l --only-use-input-alleles\n",
      "                   Only provide variant calls and genotype likelihoods for sites\n",
      "                   and alleles which are provided in the VCF input, and provide\n",
      "                   output in the VCF for all input alleles, not just those which\n",
      "                   have support in the data.\n",
      "   --haplotype-basis-alleles VCF\n",
      "                   When specified, only variant alleles provided in this input\n",
      "                   VCF will be used for the construction of complex or haplotype\n",
      "                   alleles.\n",
      "   --report-all-haplotype-alleles\n",
      "                   At sites where genotypes are made over haplotype alleles,\n",
      "                   provide information about all alleles in output, not only\n",
      "                   those which are called.\n",
      "   --report-monomorphic\n",
      "                   Report even loci which appear to be monomorphic, and report all\n",
      "                   considered alleles, even those which are not in called genotypes.\n",
      "                   Loci which do not have any potential alternates have '.' for ALT.\n",
      "\n",
      "reporting:\n",
      "\n",
      "   -P --pvar N     Report sites if the probability that there is a polymorphism\n",
      "                   at the site is greater than N.  default: 0.0.  Note that post-\n",
      "                   filtering is generally recommended over the use of this parameter.\n",
      "\n",
      "population model:\n",
      "\n",
      "   -T --theta N    The expected mutation rate or pairwise nucleotide diversity\n",
      "                   among the population under analysis.  This serves as the\n",
      "                   single parameter to the Ewens Sampling Formula prior model\n",
      "                   default: 0.001\n",
      "   -p --ploidy N   Sets the default ploidy for the analysis to N.  default: 2\n",
      "   -J --pooled-discrete\n",
      "                   Assume that samples result from pooled sequencing.\n",
      "                   Model pooled samples using discrete genotypes across pools.\n",
      "                   When using this flag, set --ploidy to the number of\n",
      "                   alleles in each sample or use the --cnv-map to define\n",
      "                   per-sample ploidy.\n",
      "   -K --pooled-continuous\n",
      "                   Output all alleles which pass input filters, regardles of\n",
      "                   genotyping outcome or model.\n",
      "\n",
      "reference allele:\n",
      "\n",
      "   -Z --use-reference-allele\n",
      "                   This flag includes the reference allele in the analysis as\n",
      "                   if it is another sample from the same population.\n",
      "   --reference-quality MQ,BQ\n",
      "                   Assign mapping quality of MQ to the reference allele at each\n",
      "                   site and base quality of BQ.  default: 100,60\n",
      "\n",
      "allele scope:\n",
      "\n",
      "   -I --no-snps    Ignore SNP alleles.\n",
      "   -i --no-indels  Ignore insertion and deletion alleles.\n",
      "   -X --no-mnps    Ignore multi-nuceotide polymorphisms, MNPs.\n",
      "   -u --no-complex Ignore complex events (composites of other classes).\n",
      "   -n --use-best-n-alleles N\n",
      "                   Evaluate only the best N SNP alleles, ranked by sum of\n",
      "                   supporting quality scores.  (Set to 0 to use all; default: all)\n",
      "   -E --max-complex-gap N\n",
      "      --haplotype-length N\n",
      "                   Allow haplotype calls with contiguous embedded matches of up\n",
      "                   to this length.  (default: 3)\n",
      "   --min-repeat-size N\n",
      "                   When assembling observations across repeats, require the total repeat\n",
      "                   length at least this many bp.  (default: 5)\n",
      "   --min-repeat-entropy N\n",
      "                   To detect interrupted repeats, build across sequence until it has\n",
      "                   entropy > N bits per bp.  (default: 0, off)\n",
      "   --no-partial-observations\n",
      "                   Exclude observations which do not fully span the dynamically-determined\n",
      "                   detection window.  (default, use all observations, dividing partial\n",
      "                   support across matching haplotypes when generating haplotypes.)\n",
      "\n",
      "indel realignment:\n",
      "\n",
      "   -O --dont-left-align-indels\n",
      "                   Turn off left-alignment of indels, which is enabled by default.\n",
      "\n",
      "input filters:\n",
      "\n",
      "   -4 --use-duplicate-reads\n",
      "                   Include duplicate-marked alignments in the analysis.\n",
      "                   default: exclude duplicates marked as such in alignments\n",
      "   -m --min-mapping-quality Q\n",
      "                   Exclude alignments from analysis if they have a mapping\n",
      "                   quality less than Q.  default: 1\n",
      "   -q --min-base-quality Q\n",
      "                   Exclude alleles from analysis if their supporting base\n",
      "                   quality is less than Q.  default: 0\n",
      "   -R --min-supporting-allele-qsum Q\n",
      "                   Consider any allele in which the sum of qualities of supporting\n",
      "                   observations is at least Q.  default: 0\n",
      "   -Y --min-supporting-mapping-qsum Q\n",
      "                   Consider any allele in which and the sum of mapping qualities of\n",
      "                   supporting reads is at least Q.  default: 0\n",
      "   -Q --mismatch-base-quality-threshold Q\n",
      "                   Count mismatches toward --read-mismatch-limit if the base\n",
      "                   quality of the mismatch is >= Q.  default: 10\n",
      "   -U --read-mismatch-limit N\n",
      "                   Exclude reads with more than N mismatches where each mismatch\n",
      "                   has base quality >= mismatch-base-quality-threshold.\n",
      "                   default: ~unbounded\n",
      "   -z --read-max-mismatch-fraction N\n",
      "                   Exclude reads with more than N [0,1] fraction of mismatches where\n",
      "                   each mismatch has base quality >= mismatch-base-quality-threshold\n",
      "                   default: 1.0\n",
      "   -$ --read-snp-limit N\n",
      "                   Exclude reads with more than N base mismatches, ignoring gaps\n",
      "                   with quality >= mismatch-base-quality-threshold.\n",
      "                   default: ~unbounded\n",
      "   -e --read-indel-limit N\n",
      "                   Exclude reads with more than N separate gaps.\n",
      "                   default: ~unbounded\n",
      "   -0 --standard-filters  Use stringent input base and mapping quality filters\n",
      "                   Equivalent to -m 30 -q 20 -R 0 -S 0\n",
      "   -F --min-alternate-fraction N\n",
      "                   Require at least this fraction of observations supporting\n",
      "                   an alternate allele within a single individual in the\n",
      "                   in order to evaluate the position.  default: 0.2\n",
      "   -C --min-alternate-count N\n",
      "                   Require at least this count of observations supporting\n",
      "                   an alternate allele within a single individual in order\n",
      "                   to evaluate the position.  default: 2\n",
      "   -3 --min-alternate-qsum N\n",
      "                   Require at least this sum of quality of observations supporting\n",
      "                   an alternate allele within a single individual in order\n",
      "                   to evaluate the position.  default: 0\n",
      "   -G --min-alternate-total N\n",
      "                   Require at least this count of observations supporting\n",
      "                   an alternate allele within the total population in order\n",
      "                   to use the allele in analysis.  default: 1\n",
      "   -! --min-coverage N\n",
      "                   Require at least this coverage to process a site.  default: 0\n",
      "\n",
      "population priors:\n",
      "\n",
      "   -k --no-population-priors\n",
      "                   Equivalent to --pooled-discrete --hwe-priors-off and removal of\n",
      "                   Ewens Sampling Formula component of priors.\n",
      "\n",
      "mappability priors:\n",
      "\n",
      "   -w --hwe-priors-off\n",
      "                   Disable estimation of the probability of the combination\n",
      "                   arising under HWE given the allele frequency as estimated\n",
      "                   by observation frequency.\n",
      "   -V --binomial-obs-priors-off\n",
      "                   Disable incorporation of prior expectations about observations.\n",
      "                   Uses read placement probability, strand balance probability,\n",
      "                   and read position (5'-3') probability.\n",
      "   -a --allele-balance-priors-off\n",
      "                   Disable use of aggregate probability of observation balance between alleles\n",
      "                   as a component of the priors.\n",
      "\n",
      "genotype likelihoods:\n",
      "\n",
      "   --observation-bias FILE\n",
      "                   Read length-dependent allele observation biases from FILE.\n",
      "                   The format is [length] [alignment efficiency relative to reference]\n",
      "                   where the efficiency is 1 if there is no relative observation bias.\n",
      "   --base-quality-cap Q\n",
      "                   Limit estimated observation quality by capping base quality at Q.\n",
      "   --prob-contamination F\n",
      "                   An estimate of contamination to use for all samples.  default: 10e-9\n",
      "   --legacy-gls    Use legacy (polybayes equivalent) genotype likelihood calculations\n",
      "   --contamination-estimates FILE\n",
      "                   A file containing per-sample estimates of contamination, such as\n",
      "                   those generated by VerifyBamID.  The format should be:\n",
      "                       sample p(read=R|genotype=AR) p(read=A|genotype=AA)\n",
      "                   Sample '*' can be used to set default contamination estimates.\n",
      "\n",
      "algorithmic features:\n",
      "\n",
      "   --report-genotype-likelihood-max\n",
      "                   Report genotypes using the maximum-likelihood estimate provided\n",
      "                   from genotype likelihoods.\n",
      "   -B --genotyping-max-iterations N\n",
      "                   Iterate no more than N times during genotyping step. default: 1000.\n",
      "   --genotyping-max-banddepth N\n",
      "                   Integrate no deeper than the Nth best genotype by likelihood when\n",
      "                   genotyping. default: 6.\n",
      "   -W --posterior-integration-limits N,M\n",
      "                   Integrate all genotype combinations in our posterior space\n",
      "                   which include no more than N samples with their Mth best\n",
      "                   data likelihood. default: 1,3.\n",
      "   -N --exclude-unobserved-genotypes\n",
      "                   Skip sample genotypings for which the sample has no supporting reads.\n",
      "   -S --genotype-variant-threshold N\n",
      "                   Limit posterior integration to samples where the second-best\n",
      "                   genotype likelihood is no more than log(N) from the highest\n",
      "                   genotype likelihood for the sample.  default: ~unbounded\n",
      "   -j --use-mapping-quality\n",
      "                   Use mapping quality of alleles when calculating data likelihoods.\n",
      "   -H --harmonic-indel-quality\n",
      "                   Use a weighted sum of base qualities around an indel, scaled by the\n",
      "                   distance from the indel.  By default use a minimum BQ in flanking sequence.\n",
      "   -D --read-dependence-factor N\n",
      "                   Incorporate non-independence of reads by scaling successive\n",
      "                   observations by this factor during data likelihood\n",
      "                   calculations.  default: 0.9\n",
      "   -= --genotype-qualities\n",
      "                   Calculate the marginal probability of genotypes and report as GQ in\n",
      "                   each sample field in the VCF output.\n",
      "\n",
      "debugging:\n",
      "\n",
      "   -d --debug      Print debugging output.\n",
      "   -dd             Print more verbose debugging output (requires \"make DEBUG\")\n",
      "\n",
      "\n",
      "author:   Erik Garrison <erik.garrison@bc.edu>, Marth Lab, Boston College, 2010-2014\n",
      "version:  v0.9.21\n"
     ]
    }
   ],
   "source": [
    "## freebayes is good variant caller but it takes BAM file\n",
    "####to do this conversion we can use samtools\n",
    "! ~/miniconda3/bin/freebayes --help\n",
    "! ~/miniconda3/bin/samtools view -S -b /mnt/d/variant_call_example/Coriell-NA12877.sam > /mnt/d/variant_call_example/Coriell-NA12877.bam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "proprietary-satellite",
   "metadata": {},
   "outputs": [],
   "source": [
    "! ~/miniconda3/bin/samtools sort /mnt/d/variant_call_example/Coriell-NA12877.bam -o /mnt/d/variant_call_example/Coriell-NA12877-sorted.bam\n",
    "! ~/miniconda3/bin/samtools index /mnt/d/variant_call_example/Coriell-NA12877-sorted.bam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "floppy-particle",
   "metadata": {},
   "outputs": [],
   "source": [
    "! ~/miniconda3/bin/freebayes -f /mnt/d/genomes/Homo_sapiens/UCSC/hg38/Sequence/Bowtie2Index/genome.fa /mnt/d/variant_call_example/Coriell-NA12877-sorted.bam > /mnt/d/variant_call_example/Coriell-NA12877.vcf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "compatible-annual",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Coriell-NA12877.bam\t\t\t   Coriell-NA12877_R2_fastqc.html\n",
      "Coriell-NA12877_R1.fastq\t\t   Coriell-NA12877_R2_fastqc.zip\n",
      "Coriell-NA12877_R1_fastqc\t\t   Coriell-NA12877.sam\n",
      "Coriell-NA12877_R1_fastqc.html\t\t   Coriell-NA12877-sorted.bam\n",
      "Coriell-NA12877_R1_fastqc_image_files.txt  Coriell-NA12877-sorted.bam.bai\n",
      "Coriell-NA12877_R1_fastqc.zip\t\t   Coriell-NA12877.vcf\n",
      "Coriell-NA12877_R2.fastq\t\t   quality_check_demo.ipynb\n",
      "Coriell-NA12877_R2_fastqc\n"
     ]
    }
   ],
   "source": [
    "## the output of freebayes is a variant call file (*.vcf) - Coriell-NA12877.vcf\n",
    "####for more on vcf see https://gatk.broadinstitute.org/hc/en-us/articles/360035531692-VCF-Variant-Call-Format\n",
    "\n",
    "! ls /mnt/d/variant_call_example"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "atomic-copying",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#CHROM\tPOS\tID\tREF\tALT\tQUAL\n",
      "chr1\t162807214\t.\tA\tC\t2.69375\n",
      "chr1\t162807221\t.\tG\tC\t2.69112\n",
      "chr1\t162807242\t.\tT\tC\t2.69961\n",
      "chr3\t130098396\t.\tC\tG\t2.70097\n",
      "chr4\t127812999\t.\tT\tG\t43.8752\n",
      "chr4\t127813032\t.\tA\tG\t29.7786\n",
      "chr9\t138139684\t.\tA\tG\t0.848001\n",
      "chr9\t138139725\t.\tT\tC\t2.61557\n",
      "chr9\t138139738\t.\tG\tA\t2.6488\n",
      "chr9\t138139772\t.\tTGGT\tGGGC\t0.384476\n",
      "chr9\t138139943\t.\tC\tG\t0.634019\n",
      "chr9\t138139978\t.\tA\tG\t3.13129\n",
      "chr10\t6804696\t.\tG\tT\t1.97689\n",
      "chr10\t28899339\t.\tT\tA\t0.401488\n",
      "chr10\t28899362\t.\tG\tT\t0.401488\n"
     ]
    }
   ],
   "source": [
    "## briefly, the vcf file will tell us genomic position along with gentoype\n",
    "####differences between sample and reference and quality of the variant call (QUAL),\n",
    "####where QUAL is defined as -10log_base10(probability that a variant does not exist)\n",
    "####the higher the QUAL, the more confident we are about the variant call\n",
    "####as a downstream process, we can choose to filter out the low QUAL variant\n",
    "####calls in a vcf\n",
    "## we can also run the applications such as SNPeff to get annotations on\n",
    "####the variant calls \n",
    "\n",
    "! sed -n 56,71p /mnt/d/variant_call_example/Coriell-NA12877.vcf | cut -f1-6"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "voluntary-clinic",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--vcf /mnt/d/variant_call_example/Coriell-NA12877.vcf\n",
      "\t--recode-INFO-all\n",
      "\t--minQ 40\n",
      "\t--out /mnt/d/variant_call_example/Coriell-NA12877-QUAL40\n",
      "\t--recode\n",
      "\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AF,Number=A,Type=Float,Description=\"Estimated allele frequency in the range (0,1]\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RO,Number=1,Type=Integer,Description=\"Reference allele observation count, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AO,Number=A,Type=Integer,Description=\"Alternate allele observations, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PRO,Number=1,Type=Float,Description=\"Reference allele observation count, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PAO,Number=A,Type=Float,Description=\"Alternate allele observations, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SRP,Number=1,Type=Float,Description=\"Strand balance probability for the reference allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SRF and SRR given E(SRF/SRR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SAP,Number=A,Type=Float,Description=\"Strand balance probability for the alternate allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SAF and SAR given E(SAF/SAR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AB,Number=A,Type=Float,Description=\"Allele balance at heterozygous sites: a number between 0 and 1 representing the ratio of reads showing the reference allele to all reads, considering only reads from individuals called as heterozygous\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=ABP,Number=A,Type=Float,Description=\"Allele balance probability at heterozygous sites: Phred-scaled upper-bounds estimate of the probability of observing the deviation between ABR and ABA given E(ABR/ABA) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPP,Number=A,Type=Float,Description=\"Read Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPPR,Number=1,Type=Float,Description=\"Read Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPP,Number=A,Type=Float,Description=\"End Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPPR,Number=1,Type=Float,Description=\"End Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=CIGAR,Number=A,Type=String,Description=\"The extended CIGAR representation of each alternate allele, with the exception that '=' is replaced by 'M' to ease VCF parsing.  Note that INDEL alleles do not have the first matched base (which is provided by default, per the spec) referred to by the CIGAR.\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GQ,Number=1,Type=Float,Description=\"Genotype Quality, the Phred-scaled marginal (or unconditional) probability of the called genotype\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GL,Number=G,Type=Float,Description=\"Genotype Likelihood, log10-scaled likelihoods of the data given the called genotype for each possible genotype generated from the reference and alternate alleles given the sample ploidy\">\n",
      "After filtering, kept 1 out of 1 Individuals\n",
      "Outputting VCF file...\n",
      "After filtering, kept 1180 out of a possible 1789 Sites\n",
      "Run Time = 2.00 seconds\n"
     ]
    }
   ],
   "source": [
    "## here we can us vcftools to extract those variant calls with a\n",
    "####QUAL of 40 or greater and save as Coriell-NA12877-QUAL40.recode.vcf\n",
    "\n",
    "! vcftools --vcf /mnt/d/variant_call_example/Coriell-NA12877.vcf --minQ 40 --recode --recode-INFO-all --out /mnt/d/variant_call_example/Coriell-NA12877-QUAL40"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "revised-budget",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Coriell-NA12877.bam\t\t\t   Coriell-NA12877_R2.fastq\n",
      "Coriell-NA12877-QUAL40.log\t\t   Coriell-NA12877_R2_fastqc\n",
      "Coriell-NA12877-QUAL40.recode.vcf\t   Coriell-NA12877_R2_fastqc.html\n",
      "Coriell-NA12877-QUAL40.vcf.log\t\t   Coriell-NA12877_R2_fastqc.zip\n",
      "Coriell-NA12877_R1.fastq\t\t   Coriell-NA12877.sam\n",
      "Coriell-NA12877_R1_fastqc\t\t   Coriell-NA12877-sorted.bam\n",
      "Coriell-NA12877_R1_fastqc.html\t\t   Coriell-NA12877-sorted.bam.bai\n",
      "Coriell-NA12877_R1_fastqc_image_files.txt  Coriell-NA12877.vcf\n",
      "Coriell-NA12877_R1_fastqc.zip\t\t   quality_check_demo.ipynb\n"
     ]
    }
   ],
   "source": [
    "! ls /mnt/d/variant_call_example/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "normal-solid",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#CHROM\tPOS\tID\tREF\tALT\tQUAL\n",
      "chr4\t127812999\t.\tT\tG\t43.8752\n",
      "chr16\t29609\t.\tC\tG\t2196.38\n",
      "chr16\t34170\t.\tC\tT\t6532.29\n",
      "chr16\t34870\t.\tC\tG\t842.8\n",
      "chr16\t35667\t.\tC\tA\t2671.02\n",
      "chr16\t58586\t.\tT\tC\t284.73\n",
      "chr16\t62515\t.\tT\tC\t57.3966\n",
      "chr16\t78180\t.\tT\tC\t231.236\n",
      "chr16\t80894\t.\tG\tC\t56.6956\n",
      "chr16\t81553\t.\tCTG\tCG\t52.9304\n",
      "chr16\t81566\t.\tC\tA\t56.6956\n",
      "chr16\t83614\t.\tT\tA\t40.2077\n",
      "chr16\t86143\t.\tC\tT\t51.376\n",
      "chr16\t92283\t.\tG\tT\t54.3723\n",
      "chr16\t98354\t.\tG\tA\t57.3966\n"
     ]
    }
   ],
   "source": [
    "! sed -n 56,71p /mnt/d/variant_call_example/Coriell-NA12877-QUAL40.recode.vcf | cut -f1-6"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "adjustable-steam",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.8.5 64-bit ('base': conda)",
   "language": "python",
   "name": "python38564bitbasecondad60fe12c1b1046abb4ad55e6f4a5c2b3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
